libdrgn/arch_s390.c

```diff
@@ -0,0 +1,214 @@
+			uint64_t r_offset, uint32_t r_type, const int64_t *r_addend,
+			uint64_t sym_value)
+{
+	struct drgn_error *err = NULL;
+
+	switch (r_type) {
+	case R_390_NONE:
+		break;
+	case R_390_8:		/* Direct 8 bit.   */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend, 0, 8, 0);
+		break;
+	case R_390_12:		/* Direct 12 bit.  */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend, 0, 12, 0);
+		break;
+	case R_390_16:		/* Direct 16 bit.  */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend, 0, 16, 0);
+		break;
+	case R_390_20:		/* Direct 20 bit.  */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend, 1, 20, 0);
+		break;
+	case R_390_32:		/* Direct 32 bit.  */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend, 0, 32, 0);
+		break;
+	case R_390_64:		/* Direct 64 bit.  */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend, 0, 64, 0);
+		break;
+	case R_390_PC16:	/* PC relative 16 bit.  */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend - r_offset, 1, 16, 0);
+		break;
+	case R_390_PC16DBL:	/* PC relative 16 bit shifted by 1.  */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend - r_offset, 1, 16, 1);
+		break;
+	case R_390_PC32DBL:	/* PC relative 32 bit shifted by 1.  */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend - r_offset, 1, 32, 1);
+		break;
+	case R_390_PC32:	/* PC relative 32 bit.  */
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend - r_offset, 1, 32, 0);
+		break;
+	case R_390_PC64:	/* PC relative 64 bit.	*/
+		err = apply_rela_bits(relocating->buf, r_offset, *r_addend - r_offset, 1, 64, 0);
+		break;
+	default:
+		err = DRGN_UNKNOWN_RELOCATION_TYPE(r_type);
+		break;
+	}
+	return err;
+}
+
+static struct drgn_error *
+fallback_unwind_s390(struct drgn_program *prog,
+		     struct drgn_register_state *regs,
+		     struct drgn_register_state **ret)
+{
+	struct optional_uint64 bc = drgn_register_state_get_u64(prog, regs, bc);
+	struct drgn_qualified_type frame_type;
+	struct drgn_type_member *member;
+	struct drgn_error *err;
+	uint64_t bit_offset, nextbc;
+	uint8_t frame[256];
+
+	if (!bc.has_value)
+		return &drgn_stop;
+
+	err = drgn_program_read_memory(prog, frame, bc.value, sizeof(frame), false);
+	if (err)
+		return err;
+
+	err = drgn_program_find_type(prog, "struct stack_frame", NULL, &frame_type);
+	if (err)
+		return err;
+
+	err = drgn_type_find_member(frame_type.type, "gprs", &member, &bit_offset);
+	if (err)
+		return err;
+
+	drgn_register_state_set_range_from_buffer(*ret, r6, r15, frame + bit_offset / 8);
+	drgn_register_state_set_pc_from_register(prog, *ret, r14);
+	err = drgn_type_find_member(frame_type.type, "back_chain", &member, &bit_offset);
+	if (err)
+		return err;
+
+	nextbc = *(uint64_t *)(frame + bit_offset / 8);
+	if (!nextbc)
+		return &drgn_stop;
+	drgn_register_state_set_from_u64(prog, *ret, bc, nextbc);
+	return NULL;
+}
+
+static struct drgn_error *
```

Comment:
Could you please add an explanation and reference to the corresponding kernel code? See the AArch64 version for an example: https://github.com/osandov/drgn/blob/330c71b5b59906c1336df9f2de19257c20f749ab/libdrgn/arch_aarch64.c#L194-L199

Reply:
On s390, we allocate a struct pt_regs on the stack when we enter the kernel. I haven't added the code to handle that yet, but will do when i submit the next version. The current code only handles the stack frame entries written when a function is called.